rust/kernel/clk.rs

```diff
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Clock subsystem
+//!
+//! C header: [`include/linux/clk.h`](../../../../include/linux/clk.h)
+
+use crate::{
+    bindings,
+    device::Device,
+    error::{from_kernel_err_ptr, Error, Result},
+    str::CStr,
+};
+
+/// Represents `struct clk *`
+pub struct Clk(*mut bindings::clk);
+
+impl Clk {
+    /// returns value of rate field of `struct clk`
+    pub fn get_rate(&self) -> u32 {
+        // SAFETY: self object contains aquired, valid pointer to `struct clk`
+        unsafe { bindings::clk_get_rate((*self).0) }
+    }
+}
+
+/// clk_prepare_enable helps cases using clk_enable in non-atomic context
+pub fn prepare_enable(clk: &mut Clk) -> Result {
+    // SAFETY: clk is an invariant containing non-null pointer to `struct clk`
```

Comment:
Note that non-NULL is not necessarily valid (I assume `clk_prepare_enable` needs it to be valid, not just non-NULL). I would simply use the same comment as in the others, unless there is something else going on.

Reply:
Yes, it is typical use-case so the standard comment `clk is valid by the type invariant.` is probably fitting.